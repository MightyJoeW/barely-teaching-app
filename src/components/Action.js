import React from "react";

const Action = props => (
  <div>
    <p>
      <button
        className="big-button"
        onClick={props.handlePick}
        disabled={!props.hasOptions}
      >
        What should I do?{" "}
      </button>
    </p>
  </div>
);

export default Action;
