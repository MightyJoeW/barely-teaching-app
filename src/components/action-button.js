// EXTERNAL DEPENDENCIES
import React from "react";

// COMPONENT DEFINITION
const ActionButton = props => (
  <button
    className="big-button"
    onClick={props.handlePick}
    disabled={!props.hasOptions}
  >
    What should I do?
  </button>
);

export default ActionButton;
