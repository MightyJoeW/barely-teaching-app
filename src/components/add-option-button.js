// EXTERNAL DEPENDENCIES
import React, { Fragment } from "react";

// COMPONENT DEFINITION
export default class AddOptionButton extends React.Component {
  state = {
    error: undefined
  };

  handleAddOptionButton = e => {
    e.preventDefault();

    const option = e.target.elements.option.value.trim(); // .trim removes whitespace
    const error = this.props.handleAddOptionButton(option);

    this.setState(() => ({ error }));

    // if no error, clear input
    if (!error) {
      e.target.elements.option.value = "";
    }
  };

  render() {
    const { error } = this.state;
    return (
      <Fragment>
        {error && <p className="add-option-error">{error}</p>}
        <form className="add-option" onSubmit={this.handleAddOptionButton}>
          <input className="add-option__input" type="text" name="option" />
          <button className="button"> Add Option </button>
        </form>
      </Fragment>
    );
  }
}
