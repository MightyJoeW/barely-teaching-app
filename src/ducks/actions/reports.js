import uuid from "uuid";
import database from "../../firebase/firebase";

// ADD_REPORT
export const addReport = report => ({
  type: "ADD_REPORT",
  report
});

export const startAddReport = (reportData = {}) => {
  return dispatch => {
    const {
      student_name = "",
      note = "",
      period = 0,
      createdAt = 0
    } = reportData;
    const report = { student_name, note, period, createdAt };

    return database
      .ref("reports")
      .push(report)
      .then(ref => {
        dispatch(
          addReport({
            id: ref.key,
            ...report
          })
        );
      });
  };
};

// REMOVE_REPORT
export const removeReport = ({ id } = {}) => ({
  type: "REMOVE_REPORT",
  id
});

export const startRemoveReport = ({ id } = {}) => {
  return dispatch => {
    return database
      .ref(`reports/${id}`)
      .remove()
      .then(() => {
        dispatch(removeReport({ id }));
      });
  };
};

// EDIT_REPORT
export const editReport = (id, updates) => ({
  type: "EDIT_REPORT",
  id,
  updates
});

export const startEditReport = (id, updates) => {
  return dispatch => {
    return database
      .ref(`reports/${id}`)
      .update(updates)
      .then(() => {
        dispatch(editReport(id, updates));
      });
  };
};

// SET_REPORTS
export const setReports = reports => ({
  type: "SET_REPORTS",
  reports
});

export const startSetReports = () => {
  return dispatch => {
    return database
      .ref("reports")
      .once("value")
      .then(snapshot => {
        const reports = [];

        snapshot.forEach(childSnapshot => {
          reports.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          });
        });

        dispatch(setReports(reports));
      });
  };
};
