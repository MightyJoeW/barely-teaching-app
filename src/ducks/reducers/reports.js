// Reports Reducer

const reportsReducerDefaultState = [];

export default (state = reportsReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_REPORT":
      return [...state, action.report];
    case "REMOVE_REPORT":
      return state.filter(({ id }) => id !== action.id);
    case "EDIT_REPORT":
      return state.map(report => {
        if (report.id === action.id) {
          return { ...report, ...action.updates };
        } else {
          return report;
        }
      });
    default:
      return state;
  }
};
