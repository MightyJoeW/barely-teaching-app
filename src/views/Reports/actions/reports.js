import uuid from "uuid";

// ADD_REPORT
export const addReport = (
  { student_name = "", note = "", period = 0, createdAt = 0 } = {}
) => ({
  type: "ADD_REPORT",
  report: {
    id: uuid(),
    student_name,
    note,
    period,
    createdAt
  }
});

// REMOVE_REPORT
export const removeReport = ({ id } = {}) => ({
  type: "REMOVE_REPORT",
  id
});

// EDIT_REPORT
export const editReport = (id, updates) => ({
  type: "EDIT_REPORT",
  id,
  updates
});
