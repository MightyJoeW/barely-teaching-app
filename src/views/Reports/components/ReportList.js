import React from "react";
import { connect } from "react-redux";
import ReportListItem from "./ReportListItem";
import selectReports from "../selectors/reports";

export const ReportList = props => (
  <div>
    {props.reports.length === 0 ? (
      <p>No reports</p>
    ) : (
      props.reports.map(report => {
        return <ReportListItem key={report.id} {...report} />;
      })
    )}
  </div>
);

const mapStateToProps = state => {
  return {
    reports: selectReports(state.reports, state.filters)
  };
};

export default connect(mapStateToProps)(ReportList);
