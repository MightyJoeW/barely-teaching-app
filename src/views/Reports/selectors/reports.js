import moment from "moment";

// Get visible reports
export default (reports, { text, sortBy, startDate, endDate }) => {
  return reports
    .filter(report => {
      const createdAtMoment = moment(report.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, "day")
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, "day")
        : true;
      const textMatch = report.student_name
        .toLowerCase()
        .includes(text.toLowerCase());

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return a.createdAt < b.createdAt ? 1 : -1;
      } else if (sortBy === "period") {
        return a.period < b.period ? 1 : -1;
      }
    });
};
