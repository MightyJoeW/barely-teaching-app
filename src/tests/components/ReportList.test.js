import React from "react";
import { shallow } from "enzyme";
import { ReportList } from "../../views/Reports/components/ReportList";
import reports from "../fixtures/reports";

test("should render ReportList with reports", () => {
  const wrapper = shallow(<ReportList reports={reports} />);
  expect(wrapper).toMatchSnapshot();
});

test("should render ReportList with empty message", () => {
  const wrapper = shallow(<ReportList reports={[]} />);
  expect(wrapper).toMatchSnapshot();
});
