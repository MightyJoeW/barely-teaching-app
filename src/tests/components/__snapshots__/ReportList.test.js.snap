// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ReportList with empty message 1`] = `
<div>
  <p>
    No reports
  </p>
</div>
`;

exports[`should render ReportList with reports 1`] = `
<div>
  <ReportListItem
    createdAt={0}
    id="1"
    key="1"
    note=""
    period={1}
    student_name="Jack"
  />
  <ReportListItem
    createdAt={-345600000}
    id="2"
    key="2"
    note=""
    period={4}
    student_name="Charlie"
  />
  <ReportListItem
    createdAt={345600000}
    id="3"
    key="3"
    note=""
    period={3}
    student_name="Fred"
  />
</div>
`;
